server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java

```diff
@@ -94,6 +94,10 @@ public String indexMetaBlobId(SnapshotId snapshotId, IndexId indexId) {
      */
     public IndexMetaDataGenerations withAddedSnapshot(SnapshotId snapshotId, Map<IndexId, String> newLookup,
                                                       Map<String, String> newIdentifiers) {
+        final Map<String, String> identifierDeduplicator = new HashMap<>(this.identifiers.size());
```

Comment:
This is a little brute force but I figured it was the easiest change I can make here to fix the problem. There's some possible ways to optimize this by a lot more but I didn't see the point in investing too much time here for now when this fixes things just as well.